### DISCUSSION TITLE
What EccCurves does GitHub.com support?

### DISCUSSION THREAD

**Discussion by TheOnlyWei** [community/community#61683] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

What EccCurves does GitHub.com support? Recently, on 7/18/2023, we get this error on Windows server 2019 datacenter when trying to access `https://github.com/`:
```
The request was aborted: Could not create SSL/TLS secure channel
```

Repro 
1. Set EccCurves to NistP384 in windows Group Policy.
https://learn.microsoft.com/en-us/windows/win32/secauthn/tls-elliptic-curves-in-windows-10-1607-and-later#enabling-elliptic-curves
2. Restart.
3. Run the following in PowerShell:
```
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
$HTTP_Request = [System.Net.WebRequest]::Create('https://github.com/')
$HTTP_Request.GetResponse()
```
It seems GitHub doesn't support NistP384 anymore, but I cannot find any blog post, changelog, or documentation about this.

**Comment by nirav-gajera** [comment]
According to the GitHub documentation, GitHub.com supports the following ECC curves:

* secp256r1
* secp384r1
* ecdsa_secp256k1

The error message you are seeing indicates that your Windows Server 2019 is configured to use NistP384, which is not supported by GitHub.com. You can either change the ECC curve in your Windows Group Policy, or you can use a different ECC curve in your PowerShell script.

Here are the steps to change the ECC curve in your Windows Group Policy:

1. Open the Group Policy Management Editor.
2. Navigate to Computer Configuration > Administrative Templates > System > Internet Communication Settings > TLS 1.2 Settings.
3. Double-click on the "Select the ECC curves for TLS 1.2" policy.
4. In the "Enabled" list, select the ECC curves that you want to use.
5. Click on "OK" to save your changes.

Once you have changed the ECC curve in your Windows Group Policy, you will need to restart your server. After your server has restarted, you should be able to access GitHub.com without any errors.

If you do not want to change the ECC curve in your Windows Group Policy, you can use a different ECC curve in your PowerShell script. The following code will use the secp256r1 ECC curve:

```
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
$HTTP_Request = [System.Net.WebRequest]::Create('https://github.com/')
$HTTP_Request.ClientCertificateOptions = [System.Net.Security.SslClientCertificateOptions]::None
$HTTP_Request.ServerCertificateValidationCallback = { $true }
$HTTP_Request.GetResponse()
```

This code will bypass the ECC curve validation and allow you to access GitHub

**Reply by TheOnlyWei** [reply]
@nirav-gajera Thanks for the reply, but I am not sure if that list is accurate or the full list. This is the default ECC curve order for Windows server 2019 datacenter as listed in my group policy "Help" section for "ECC Curve Order" that works with GitHub:
curve25519
NistP256
NistP384

Obviously, GitHub no longer supports NistP384 as of around 7/18/2023 since it seems to have worked in the past. But that means the first two ECC curves worked against GitHub, which would be on top of what you listed. Do you have an official GitHub document link where you got the list from?

**Reply by nirav-gajera** [reply]
I did not find an official GitHub document that lists the supported ECC curves. However, I found the list on the following GitHub repositories:

* RustCrypto/elliptic-curves: https://github.com/RustCrypto/elliptic-curves
* ANSSI-FR/libecc: https://github.com/ANSSI-FR/libecc
* nakov/Practical-Cryptography-for-Developers-Book: https://github.com/nakov/Practical-Cryptography-for-Developers-Book
